pkg/proxy/nftables/proxier.go

```diff
@@ -1639,20 +1655,31 @@ func (proxier *Proxier) writeServiceToEndpointRules(tx *knftables.Transaction, s
 			ipX = "ip6"
 		}
 
-		for _, ep := range endpoints {
-			epInfo, ok := ep.(*endpointInfo)
-			if !ok {
-				continue
-			}
+		// Rule to map an known source IP to an endpoint IP.
+		tx.Add(&knftables.Rule{
+			Chain: svcChain,
+			Rule: knftables.Concat(
+				ipX, "daddr set", ipX, "saddr map", "@", svcInfo.affinityMapToEPName,
+			),
+		})
+
+		// Rule to look up endpoint chain from an endpoint IP.
+		tx.Add(&knftables.Rule{
+			Chain: svcChain,
+			Rule: knftables.Concat(
+				ipX, "daddr vmap", "@", svcInfo.affinityMapToChainName,
```

Comment:
So the packet's dest IP will be overridden once by the first map, then DNATed once in the endpoint chain?
Intuitively it needs one vmap only, which maps the source IP to the endpoint chain directly and performs DNAT only once. I suppose there are some limitations in nftables preventing it being implemented so?

Not sure if it there is any side effect caused by the two-step DNAT, but I feel the approach looks a bit like a workaround.

@danwinship is this approach common?

Reply:
The limitation is verdict map does not support the timeout flag required to refresh the timeout of a cached client once a packet from that client passes through kube-proxy. That's why an additional map to EP IP is used here. 

I wanted to map a client to endpoint's packet mark instead of endpoint's dest IP, but setting packet mark at this rule could overwrite the packet mark set by mark-for-masquerade. Another challenge is to keep packet mark unique for each endpoint. That's why setting packet's dest IP to endpoint's IP is chosen. I am not sure updating dest IP in the service rule is the same as the DNAT rule in the endpoint chain. I've not met any issue with it so far, though .

Reply:
> Intuitively it needs one vmap only, which maps the source IP to the endpoint chain directly and performs DNAT only once.

Yeah, it seems like we ought to be able to do that. (I confess I didn't look too deeply at the rules before; was waiting for the update to `proxier_test.go` so I could look at the updated dump output.)

> Not sure if it there is any side effect caused by the two-step DNAT

Nothing outside of netfilter would ever see the intermediate daddr value, so it's not awful, but it's definitely weird, and may cause netfilter to do inefficient things?

> is this approach common?

Is _anything_ "common" in nftables yet? :grimacing: 

> I wanted to map a client to endpoint's packet mark

I want to avoid using the packet mark at all; we currently still use it for masquerading, but there's a plan in the KEP to try to get rid of that.

> I am not sure updating dest IP in the service rule is the same as the DNAT rule in the endpoint chain.

It's not; it doesn't create a DNAT conntrack entry. But in this case you're still calling `dnat` once you get to the endpoint chain, so the DNAT entry gets created there. Though I would expect that to create the _wrong_ DNAT entry such that when the reply packet gets un-DNATted, it ends up with the wrong destination address... If that's _not_ happening, it must be because the `ip daddr set` is effectively getting optimized out, but that seems like a bug and I'm not sure I'd want to rely on that.